torch/csrc/PyInterpreter.cpp

```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+  // 5 (3) our buffer is no longer size 5 (because at some point previously we
+  // saw a size > 5),
+  //     and our new size is different from the current buffer size
+  needs_resize = !is_buffer_allocated ||
+      ((!is_smallvector || new_size > 5) && curr_size != new_size);
+  if (needs_resize) {
+    auto set_to_smallvector = false;
+    // If our current buffer is not the right size (eithr because we haven't
+    // allocated it yet, or there was a metadata mutation that changed the
+    // number of dims of the tensor), allocate a fresh buffer. Note that this
+    // will trash the previous buffer if there already was one, invalidating any
+    // existing SymIntArrayRef's from an old .sym_size() call.
+    auto new_buffer_size = new_size;
+    if (!is_buffer_allocated && new_size <= 5) {
+      // This is the smallvector optimization
+      new_buffer_size = 5;
+      set_to_smallvector = true;
+    }
+    T* ptr = new T[new_buffer_size];
+    auto capsule =
+        py::capsule(ptr, [](void* p) { delete[] reinterpret_cast<T*>(p); });
+    int64_t idx = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++idx) {
+      ptr[idx] = py::cast<T>(*it);
+    }
+    // Set the buffer
+    set_tensor_attr_with_capsule(tensor, capsule, base_attr_name);
+    // Set the len buffer
+    py::handle(mb_obj.value()).attr(buffer_len_attr_name.c_str()) = new_size;
+    // Set whether or not the buffer is now using the smallvector optimization
+    // (sized to 5)
+    py::handle(mb_obj.value()).attr(buffer_issmallvector_attr_name.c_str()) =
+        set_to_smallvector;
+  } else {
+    auto curr_buffer_pyobj = py::handle(mb_obj.value()).attr(base_attr_name);
+    void* buffer_pycapsule =
+        PyCapsule_GetPointer(curr_buffer_pyobj.ptr(), nullptr);
+    auto curr_buffer = reinterpret_cast<T*>(buffer_pycapsule);
+
+    // Overwrite the buffer with our new values, but only if any of them changed
+    // (due to a metadata mutation).
+    // This is technically not thread safe (maybe we should put the GIL here?),
+    // but *only* if we actually have to update, which only happens if there's a
+    // metadata mutation.
+    int64_t idx = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++idx) {
+      auto actual_val = py::cast<T>(*it);
+      if constexpr (std::is_same_v<T, c10::SymInt>) {
+        // if our SymInts are symbolic, we are *not* doing an equality check on
+        // the symints. we just want to see if the nodes are the same. this is
+        // because we don't want to introduce any guards here.
+        if ((curr_buffer[idx].is_heap_allocated() !=
+             actual_val.is_heap_allocated()) ||
+            (!curr_buffer[idx].is_heap_allocated() &&
+             curr_buffer[idx] != actual_val) ||
+            (curr_buffer[idx].is_heap_allocated() &&
+             curr_buffer[idx].toSymNodeImplUnowned() !=
+                 actual_val.toSymNodeImplUnowned())) {
```

Comment:
Nyeh, we should probably just add a method to SymInt that does this, you literally just want to look and see if the bit pattern is the same, you really shouldn't need a conjunction of tests

Reply:
Definitely fair, will add

Reply:
Bikeshedding the name: `SymInt::identity_equals`?